ledger/complete/mtrie/flattener/decode.go

```diff
@@ -0,0 +1,147 @@
+	// minBufSize is a failsafe and is only used when len(scratch) is much smaller
+	// than expected.  len(scratch) is 4096 by default, so minBufSize isn't likely to be used.
+	const minBufSize = 1024
+
+	if len(scratch) < minBufSize {
+		scratch = make([]byte, minBufSize)
+	}
+
+	// fixLengthSize is the size of shared data of leaf node and interim node
+	const fixLengthSize = encNodeTypeSize + encHeightSize + encHashSize
+
+	_, err = io.ReadFull(reader, scratch[:fixLengthSize])
+	if err != nil {
+		return false, 0, 0, 0, ledger.DummyPath, nil, hash.DummyHash, fmt.Errorf("failed to read fixed-length part of serialized node: %w", err)
+	}
+
+	pos := 0
+
+	// Decode node type (1 byte)
+	nType := scratch[pos]
+	pos += encNodeTypeSize
+
+	if nType != byte(leafNodeType) && nType != byte(interimNodeType) {
+		return false, 0, 0, 0, ledger.DummyPath, nil, hash.DummyHash, fmt.Errorf("failed to decode node type %d", nType)
+	}
+
+	// Decode height (2 bytes)
+	height = binary.BigEndian.Uint16(scratch[pos:])
+	pos += encHeightSize
+
+	// Decode and create hash.Hash (32 bytes)
+	nodeHash, err = hash.ToHash(scratch[pos : pos+encHashSize])
+	if err != nil {
+		return false, 0, 0, 0, ledger.DummyPath, nil, hash.DummyHash, fmt.Errorf("failed to decode hash of serialized node: %w", err)
+	}
+
+	if nType == byte(leafNodeType) {
+
+		// Read path (32 bytes)
+		encPath := scratch[:encPathSize]
+		_, err := io.ReadFull(reader, encPath)
+		if err != nil {
+			return false, 0, 0, 0, ledger.DummyPath, nil, hash.DummyHash, fmt.Errorf("failed to read path of serialized node: %w", err)
+		}
+
+		// Decode and create ledger.Path.
+		path, err := ledger.ToPath(encPath)
+		if err != nil {
+			return false, 0, 0, 0, ledger.DummyPath, nil, hash.DummyHash, fmt.Errorf("failed to decode path of serialized node: %w", err)
+		}
+
+		// Read encoded payload data and create ledger.Payload.
+		payload, err := readPayloadFromReader(reader, scratch)
+		if err != nil {
+			return false, 0, 0, 0, ledger.DummyPath, nil, hash.DummyHash, fmt.Errorf("failed to read and decode payload of serialized node: %w", err)
+		}
+
+		return false, height, 0, 0, path, payload, nodeHash, nil
+	}
+
+	// Read interim node
+
+	// Read left and right child index (16 bytes)
+	_, err = io.ReadFull(reader, scratch[:encNodeIndexSize*2])
+	if err != nil {
+		return true, 0, 0, 0, ledger.DummyPath, nil, hash.DummyHash, fmt.Errorf("failed to read child index of serialized node: %w", err)
+	}
+
+	pos = 0
+
+	// Decode left child index (8 bytes)
+	lchildIndex = binary.BigEndian.Uint64(scratch[pos:])
+	pos += encNodeIndexSize
+
+	// Decode right child index (8 bytes)
+	rchildIndex = binary.BigEndian.Uint64(scratch[pos:])
+
+	return true, height, lchildIndex, rchildIndex, ledger.DummyPath, nil, nodeHash, nil
+}
+
+// DecodeTrieData reads the trie data and decode it.
+// any error returned means the file is malformated
+func DecodeTrieData(reader io.Reader, scratch []byte) (
```

Comment:
This is extracted from encoding/ReadTrie function, so that it can be reused by compare-checkpoint cmd